MvvmCross/Core/Binding/BindingContext/MvxTaskBasedBindingContext.cs

```diff
@@ -0,0 +1,209 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using MvvmCross.Binding.Binders;
+using MvvmCross.Binding.Bindings;
+using MvvmCross.Platform;
+
+namespace MvvmCross.Binding.BindingContext
+{
+    /// <summary>
+    /// OnDataContextChange executes asynchronously on a worker thread
+    /// </summary>
+    public class MvxTaskBasedBindingContext : IMvxBindingContext
+    {
+        private readonly List<Action> _delayedActions = new List<Action>();
+        private readonly List<MvxBindingContext.TargetAndBinding> _directBindings = new List<MvxBindingContext.TargetAndBinding>();
+        private readonly List<KeyValuePair<object, IList<MvxBindingContext.TargetAndBinding>>> _viewBindings = new List<KeyValuePair<object, IList<MvxBindingContext.TargetAndBinding>>>();
+        private object _dataContext;
+        private IMvxBinder _binder;
+
+        public event EventHandler DataContextChanged;
+
+        public IMvxBindingContext Init(object dataContext, object firstBindingKey, IEnumerable<MvxBindingDescription> firstBindingValue)
+        {
+            this.AddDelayedAction(firstBindingKey, firstBindingValue);
+            if (dataContext != null)
+                this.DataContext = dataContext;
+
+            return this;
+        }
+
+        public IMvxBindingContext Init(object dataContext, object firstBindingKey, string firstBindingValue)
+        {
+            this.AddDelayedAction(firstBindingKey, firstBindingValue);
+            if (dataContext != null)
+                this.DataContext = dataContext;
+
+            return this;
+        }
+
+        private void AddDelayedAction(object key, string value)
+        {
+            this._delayedActions.Add(() =>
+            {
+                var bindings = this.Binder.Bind(this.DataContext, key, value);
+                foreach (var b in bindings)
+                    this.RegisterBinding(key, b);
+            });
+        }
+
+        private void AddDelayedAction(object key, IEnumerable<MvxBindingDescription> value)
+        {
+            this._delayedActions.Add(() =>
+            {
+                var bindings = this.Binder.Bind(this.DataContext, key, value);
+                foreach (var b in bindings)
+                    this.RegisterBinding(key, b);
+            });
+        }
+
+        ~MvxTaskBasedBindingContext()
+        {
+            this.Dispose(false);
+        }
+
+        public void Dispose()
+        {
+            this.Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                this.ClearAllBindings();
```

Comment:
Well it should not happen, as the view subscriptions to viewmodel changes are weak. 
But it may happen in corner cases if Dispose() is called explicitly on views in a custom presenter.

A good workaround would be to wait for tasks to end, but Dispose is not async so we can not do that.
So code using Dispose in their custom presenter needs to wait for binding tasks to finish before calling Dispose.
This may be documented in the wiki as a change. And we may expose a public Task property to wait for.
Or the code could wait a few seconds before disposing. It should be enough as executing all bindings should always be very fast <20ms.